ion-schema/src/isl/isl_type_reference.rs

```diff
@@ -143,29 +149,82 @@ impl IslTypeRefImpl {
+                        "int" | "float" | "bool" | "decimal" | "string" | "symbol" | "blob" | "clob" | "timestamp" | "struct" | "sexp" | "list" | "text" | "lob" | "number" | "any" | "nothing" => {
+                            Ok(IslTypeRefImpl::Named(type_name.to_owned(), Nullable))
+                        }
+                        _ => {
+                            invalid_schema_error(
+                                "`nullable::` modifier is only supported for built in types",
+                            )
+                        }
+                    }
                 }
 
-                Ok(IslTypeRefImpl::Named(type_name.to_owned()))
+                // check for `$null_or` type reference
+                if value.has_annotation("$null_or") {
+                    if isl_version == IslVersion::V1_0 {
+                        return invalid_schema_error(
+                            "`$null_or::` modifier is not supported before ISL 2.0",
+                        )
+                    }
+                    return Ok(IslTypeRefImpl::Named(type_name.to_owned(), NullOr))
+                }
+
+                Ok(IslTypeRefImpl::Named(type_name.to_owned(), Nothing))
             }
             IonType::Struct => {
                 if value.is_null() {
                     return invalid_schema_error(
                         "a base or alias type reference can not be null.struct",
                     )
                 }
-                if value.annotations().any(|a| a.text() == Some("nullable")) {
+                let mut isl_type_ref_modifier = Nothing;
+
+                // check for nullable type reference
+                if value.has_annotation("nullable") {
+                    if isl_version == IslVersion::V2_0 {
+                        return invalid_schema_error(
+                            "`nullable::` modifier is not supported since ISL 2.0",
+                        )
+                    }
+                    // TODO: currently it only allows for built in types to be defined with `nullable` annotation for `document` and all other type references it returns an error
                     return invalid_schema_error(
-                        "nullable type reference is only allowed for built in types",
+                        "`nullable` modifier is only supported for built-in types",
                     )
                 }
+
+                // check for `$null_or` type reference
+                if value.has_annotation("$null_or") {
+                    if isl_version == IslVersion::V1_0 {
+                        return invalid_schema_error(
+                            "`$null_or::` modifier is not supported before ISL 2.0",
+                        )
+                    }
+                    isl_type_ref_modifier = NullOr
+                }
+
```

Comment:
If you take my suggestion above, then I think you can replace all of this with:
```suggestion
                if nullability == Nullable {
                    // TODO: currently it only allows for built in types to be defined with `nullable` annotation for `document` and all other type references it returns an error
                    return invalid_schema_error(
                        "`nullable` modifier is only supported for built-in types",
                    )
                }
```
(Beware—committing this this "suggestion" block _might_ not work the way you expect it should because it covers some "deleted lines".)